### ISSUE TITLE
sympy.floor() sometimes returns the wrong answer

### ISSUE THREAD

**Issue by billzorn** [sympy/sympy#13098]
The title says it all. If you take the floor of a number slightly less than an integer, sometimes you get that integer, even though evalf() even at default precision could tell you the number was actually less. This is with sympy 1.1.1, using Python 3.6.0 from Anaconda on Ubuntu 16.04.

Here it is with log(x, 10) slightly less than one:
```
>>> sympy.sympify('floor(log(9.9999999, 10))') # correct
0
>>> sympy.sympify('floor(log(9.99999999, 10))') # incorrect
1
>>> x = sympy.sympify('log(9.99999999, 10)')
>>> x.evalf()
0.999999999565706
>>> x.evalf().is_comparable
True
```
It seems reluctant to reproduce with other numbers known to be troublesome:
```
>>> x = sympy.floor(sympy.Rational(sympy.pi.evalf(1000)) - sympy.pi)
>>> x
floor(-pi + 660427... # a huge rational
>>> x.evalf()
sympy.core.evalf.PrecisionExhausted: Failed to distinguish the expression... # understandable
>>> x.evalf(maxn=1000)
-1.0
```
Or with log(x, 2):
```
>>> sympy.sympify('floor(log(15.999999999999999999999999999999999999999999999999999999, 2))')
3
>>> sympy.sympify('floor(log(16, 2))')
4
```
Or even different powers of 10?
```
>>> sympy.sympify('floor(log(99.99999999999999999999999, 10))') # correct?
1
>>> sympy.sympify('floor(log(999.99999, 10))') # correct
2
>>> sympy.sympify('floor(log(999.999999, 10))') # incorrect?
3
```
Strange. I guess I can work around this by calling evalf() before the floor and manually ensuring I have enough precision.

**Comment by asmeurer** [comment]
My guess is it's related to https://github.com/sympy/sympy/issues/12534.

**Comment by smichr** [comment]
There is an error in logic in `calc_integer_part`. This expression `floor(log(9.99999999, 10))` is not larger than `2**30` but it is computed without precision when `assumed_size=30`. If that 30 is changed to even 31 then the result for `floor(log(9.99999999, 10))` is computed as 0 instead of 1. [Here](https://github.com/diofant/diofant/pull/179) a decision was made to remove that routine. An alternative would be to increase precision if a precisionless result is obtained from the `assumed_size` calculation.

**Comment by dixr** [comment]
It seems there is still an error somewhere:
```
>>> sympy.floor((sympy.log(2**30-1,2)+1))
30
>>> sympy.floor((sympy.log(2**31-1,2)+1))
32
```
This is on both 1.12 and 1.13.dev. 

On 1.12 we have additionally this problem:
```
>>> int((sympy.log(2**359-1,2)+1))
359
>>> int((sympy.log(2**360-1,2)+1))
361
```
which seems to have been fixed on 1.13.dev though (see [this colab](https://colab.research.google.com/drive/1EPDf_GudFRQzJeZhp-jOhxny5Z3_IGrd?usp=sharing)).
